The Great GPU Rewrite™

fuelsoft opened the pull request:
Games should now be drawn 100% accurately, barring bugs I did not find in my testing. No shortcuts are taken anymore. The call to `redraw()` is now just a stub, it should be cleaned up in the future.

1. On the first clock of *OAM*, an array of up to 10 sprites is populated with sprites present on the line. Subsequent clocks are idled (OAM is locked, so nothing can change during the rest of the period anyway).

1. Every clock during *Pixel Transfer*, a single pixel is calculated for background, window, and (possibly multiple) sprite indices.


Also some fixes:

1. Fixed palette bug - high and low index bit were swapped.

1. Overlapping sprites will now draw correctly.

1. Sprite priority should work, but I'm not 100% sure any of the games I test with actually use it - it appears to be a rarely used feature.

fuelsoft committed:
Background now drawn progressively

fuelsoft committed:
Window fixes - should now be very accurate

fuelsoft committed:
Move sprite draw to progressive draw

fuelsoft committed:
Code cleanup

fuelsoft committed:
Cleanup and bug fixes

fuelsoft committed:
Fixed infernal palette mapping bug

fuelsoft committed:
Fixed sprite overlapping bug

fuelsoft committed:
Fix sprite bug when partially off left, cleaned up legacy render code

summertriangle-dev commented:
Mostly LGTM, but I found a couple bugs from my test game collection:

- Mario draws on top of pipes when entering them in Super Mario Land 
  <img width="506" alt="Screen Shot 2020-10-22 at 02 22 39" src="https://user-images.githubusercontent.com/7821037/96854732-af6b2900-1410-11eb-8ccc-b77fbb18ffb1.png">
  I think this is a sprite priority issue - I see that you're skipping sprite draw if the background colour is 3, but [according to the docs it should be any colour except 0](https://gbdev.gg8.se/wiki/articles/Video_Display#Byte3_-_Attributes.2FFlags:).

- Some wrong tiles in Kirby's Dreamland 
  <img width="503" alt="Screen Shot 2020-10-22 at 02 41 52" src="https://user-images.githubusercontent.com/7821037/96855466-9ca52400-1411-11eb-8f49-afe97ba370ee.png">
  Not sure about this one since it only occurs for some objects.

fuelsoft committed:
Fix priority

fuelsoft committed:
Fix double height y-flip sprite bug

fuelsoft commented:
1. Yeah I remember seeing that now that I think about it, fixed.

1. Not surprised this didn't work, double height sprites only work purely by lucky design after the rewrite. Never even thought about vertical flip. Fixed.

I'd also like to bump the version to 1.2 and make another release once this is in an acceptable state.

summertriangle-dev reviewed (approved):
Yep, looks like everything’s fixed. Fire away

fuelsoft committed:
Bump version number to 1.2.0
